Add date range filters to the search UI

mmcfarland opened the pull request:
## Overview

Adds UI elements for collecting start and end date filters and passes them along to the existing API endpoint using the pre-existing querystring parameters.  Does some light validation to ensure that
`before < after`, which would otherwise raise exceptions on the backend.

Also adds a small formatting improvement to the warning dialog for areas which are too large for WDC to handle.

Connects #1933  

### Demo
#### Added filter toggle
![screenshot from 2017-07-25 09 03 56](https://user-images.githubusercontent.com/1014341/28573330-554577f8-7118-11e7-8973-94a5f16fb926.png)

#### Filters exposed with date-picker
![screenshot from 2017-07-25 09 04 15](https://user-images.githubusercontent.com/1014341/28573333-554694e4-7118-11e7-8d3f-5da26e643458.png)

#### Validation response
![screenshot from 2017-07-25 09 04 34](https://user-images.githubusercontent.com/1014341/28573331-5545863a-7118-11e7-983f-1df86ca64a69.png)

#### Warning - before
![screenshot from 2017-07-25 09 00 59](https://user-images.githubusercontent.com/1014341/28573334-55472c92-7118-11e7-84f0-2b73621b588d.png)

#### Warning - after
![screenshot from 2017-07-25 08 16 18](https://user-images.githubusercontent.com/1014341/28573332-5545fdc2-7118-11e7-9e6d-896ca3029d03.png)

### Notes
I uncovered an issue that is visible here, but also on staging (though less pronounced), captured here: #2080.  I also made a good-faith effort to match the styling shown in screenshots, but this could use some touch-ups as part of a final style pass on the larger feature.

I had to derive a bit of UX from the screenshot listed in the issue.  My decisions on the behavior are as followed, and are open to feedback:

* 0, 1 or both dates can be filled in without a validation error.  The backend appears to support this configuration.
* The datepicker library handles making sure a date format is valid, so I only ensure that the date precedence is appropriate.  If after < before in the WDC query, it will return an exception.
* I didn't do validation on blur from the field, since that would be annoying if the user was changing the values and after changing one, had them temporarily in an invalid state.  Validation is only run on search.
* Toggling the filter keeps the date values in place, so they are there when you re-toggle.  This means they are still used, even when hidden.  This could be confusing, but in my opinion, it's not any clearer than having them disappear when you click "Hide Filters".  Or if worse if they are not used when hidden, but are still there when you un-toggle.  I'm open to "Hide Filters" just clearing them from state altogether as an alternative.
* Hitting enter in a datebox will execute the search.  At first, I didn't like that approach, but in practice, having search results displayed and going back and tweaking the date seemed like a natural workflow.  Not having to tab into the query box to enter seems appropriate in that case.  

## Testing Instructions

 * A new JS dependency was added (datepicker), so npm install and rebundle the vendor and client assets.
* Navigate to the search and execute a number of searches with and without dates, or partial dates.
* Attempt to provide bad date values to the search.
* Date range input should be reflected in the results.

ajrobbins commented:
+1 for the UX decisions described in the Notes. 

>Toggling the filter keeps the date values in place, so they are there when you re-toggle. This means they are still used, even when hidden.

It sounds like a UI element like a flag or highlight of some kind would be useful to show that a filter is activated, even when the filter is hidden. /cc @jfrankl

rajadain commented on src/mmw/js/src/data_catalog/views.js:
```diff
@@ -126,20 +131,84 @@ var FormView = Marionette.ItemView.extend({
+    },
+
+    onRender: function() {
+        $('.data-catalog-date-input').datepicker();
```
Need to import `$` before we can use it. That's causing the linting errors:

```
$ kj npm run lint

js/src/data_catalog/views.js: line 151, col 9, '$' is not defined.
js/src/data_catalog/views.js: line 170, col 27, '$' is not defined.

2 errors
```

rajadain commented on src/mmw/js/src/data_catalog/views.js:
```diff
@@ -90,13 +91,14 @@ var DataCatalogWindow = Marionette.LayoutView.extend({
                 model: new modalModels.AlertModel({
                     alertMessage: "The bounding box of the current area of " +
-                                  "interest is " + Math.round(area) + " km², " +
+                                  "interest is " + formattedArea + "&nbspkm², " +
```
Missing semicolon: `&nbspkm` → `&nbsp;km`

rajadain commented on src/mmw/js/src/data_catalog/views.js:
```diff
@@ -90,13 +91,14 @@ var DataCatalogWindow = Marionette.LayoutView.extend({
                                   "which is larger than the current maximum " +
-                                  "area of " + MAX_AREA_SQKM + " km² supported " +
-                                  "for WDC.",
+                                  "area of " + MAX_AREA_FORMATTED + "&nbspkm² " +
```
Here too.

rajadain commented on src/mmw/js/src/data_catalog/models.js:
```diff
@@ -17,6 +17,8 @@ var Catalog = Backbone.Model.extend({
         id: '',
         name: '',
         description: '',
+        dateStart: null,
```
Consider changing to `fromDate` and `toDate` to match the API.

rajadain commented on src/mmw/npm-shrinkwrap.json:
```diff
@@ -44,6 +44,11 @@
       "from": "https://registry.npmjs.org/bootstrap/-/bootstrap-3.3.4.tgz",
       "resolved": "https://registry.npmjs.org/bootstrap/-/bootstrap-3.3.4.tgz"
     },
+    "bootstrap-datepicker": {
```
👏  I like the neatness of the shrinkwrap changes! There is a typo in the commit message though `javascrip` → `javascript`.

rajadain commented:
A search with no dates is failing:

![image](https://user-images.githubusercontent.com/1430060/28580145-acd6e9ac-712c-11e7-94af-a3b149817249.png)

rajadain commented on src/mmw/js/src/data_catalog/views.js:
```diff
@@ -126,20 +131,84 @@ var FormView = Marionette.ItemView.extend({
+            after = moment(this.model.get('dateEnd'), dateFormat),
+            isValid = false;
+
+        if (!before || !after) {
```
Changing this to

```js
if (!before.date() || !after.date())
```

fixes the issue of eager validation:

![image](https://user-images.githubusercontent.com/1430060/28580207-d462aee8-712c-11e7-8f33-184ec7119762.png)

mmcfarland replied:
Looks like the validation message is behaving strangely in your screenshot, I'll revist and make sure I didn't alter something as I was cleaning up my commits.

rajadain commented on src/mmw/js/src/data_catalog/templates/form.html:
```diff
@@ -1,6 +1,35 @@
+        <input
+            class="data-catalog-date-input form-control"
+            data-date-range="start"
+            value="{{dateStart}}"
```
Before should map to `dateEnd` and After should map to `dateStart`

rajadain replied:
The current setup is wrong:

![image](https://user-images.githubusercontent.com/1430060/28581149-880b7e82-712f-11e7-9728-67b8e1768418.png)

since this describes an infinite date range with a gap in between. However, reading "After" come first and "Before" later could also be counterintuitive. Perhaps we should use other labels, like "From" and "To" or "Till".

mmcfarland replied:
Yeah, I found the labels problematic to map to the API attributes.  I'll assume they were determined casually and come up with something that can be better understood.

mmcfarland commented:
Added a number of fixups which, in addition to some minor cleanup, alters the display of the date ranges so that it reads as "Between X and Y", to prevent confusion about which end of the spectrum a particular date can be on.  This allows a more straightforward mapping from `textbox -> variable -> api`.

You can see the filters being applied in action, here:

![screenshot from 2017-07-25 13 46 08](https://user-images.githubusercontent.com/1014341/28586583-4e2d06fa-7142-11e7-8612-6ec92806649e.png)
![screenshot from 2017-07-25 13 45 47](https://user-images.githubusercontent.com/1014341/28586584-4e33dfac-7142-11e7-8d5e-5b56efafd0f7.png)
--
The validation message is updated as well:

![screenshot from 2017-07-25 14 07 16](https://user-images.githubusercontent.com/1014341/28586678-9816de8a-7142-11e7-81ad-0d780ebeb4c7.png)

jfrankl commented:
A few thoughts:

I prefer when the validation message is on its own line below the inputs. In some screen captures, it is displayed to the right of the inputs.

Would be good if there were a clear button to remove the date when one has been entered. The wireframes show a "X Reset" button after a date has been entered. An "X" button inside the inputs, on the right side, would also work.

Other than that, this looks good to me, and once it's in I can do some minor style cleanup.

rajadain commented:
Tested the new design in a smaller window, the validation message is correctly below, not beside, the input fields:

![image](https://user-images.githubusercontent.com/1430060/28590261-8b37101a-714f-11e7-995b-8f4772ccb070.png)

mmcfarland commented:
@rajadain Thanks for the review.  I squashed my fixups, and added a new commit addressing @jfrankl's comment on the "clear" buttons.  These weren't in the wireframes posted in the issue, but I interpreted it as:

![screenshot from 2017-07-25 16 02 10](https://user-images.githubusercontent.com/1014341/28591396-8fcd2dc2-7153-11e7-92f3-13ae1210e3a3.png)

They only show up when there is a value in the box, and clicking it will clear out the value.  Adding this in created some unwanted visual effects (the box grows and shrinks if they are present/absent, and it's no longer flush with the edge).  I attempted to restore my previous style, but think the effort would be better spent during the design pass.  Could you give it one more spin?

rajadain commented:
Taking another look.
